### ISSUE TITLE
How can we improve this

### ISSUE THREAD

**Issue by ThorvaldAagaard** [lorserker/ben#68]
In this position
![image](https://github.com/lorserker/ben/assets/10560652/51f763d1-f099-45f7-aae6-2b59c52b3bbd)
East needs to find s discard.
Double dummy East can see all tricks for declarer, so the discard is decided by the Neural Network

![image](https://github.com/lorserker/ben/assets/10560652/ffce20cd-59be-4e6b-81d5-8c87ecced5b3)

As KQ in hearts are missing, the HT can never be a trick, but NN suggest to drop DT

**Comment by lorserker** [comment]
we need a different (better) approach to the NN, which i don't know yet what it is.

another idea is to also do rollouts, especially in the endgame, and use that too to decide.

by rollout i mean similar algorithm like for bidding:
for every possible card: play it, then play each sample out to the end only with NN and then choose the card which works best

**Comment by ThePokerDude** [comment]
Do I understand correctly that east takes the information about it's and dummy's cards, creates the possible distributions and for each of them dubble dummy play is checked?
If yes - then for any of the options declarer taks all tricks, so for BEN it's ok to discard DT.

An option to solve it would be to choose a two step process 
1) create possible distributions for declarer hands
2) for each of these distributions we ignore the actual east's hand and for each of the possible discards create again all possible distributions and do double dummy analysis

In this case the engine would find that declarer errs some percatage of the time when HT is discarded but never when DT or DQ is discarded

**Comment by ThorvaldAagaard** [comment]
What I am looking for is a 3(4) card ending analyzer, that can help both declarer and defender, After trick 10 there will at most be 120 combinations, when you can see 2 hands. For defenders it should be not to discard card, that would allow declarer to make errors, and for declarer it is to avoid falling in the trap, that opponents does not have what hey have shown in the bidding, and use winning lines that also includes this situation.

**Comment by ThePokerDude** [comment]
There is one thing where you should be careful.
Let's say you play a NT contract and during play you found that one of the opponents did not follow suit and you know
the exact card distribution.
If you do not forward this information but just the missing cards, the solver can play for a distribution which is not possible in the actual game.

**Comment by ThorvaldAagaard** [comment]
> There is one thing where you should be careful. Let's say you play a NT contract and during play you found that one of the opponents did not follow suit and you know the exact card distribution..

This is implemented in the new play engine, and was also part of the old one

**Comment by ThePokerDude** [comment]
Yes, I was aware that this is implemented.
I just thought that if you "outsource" the game analysis to an endgame solver when a 3 card ending is reached
additionally to forwarding the information about which cards are left you need also to forward the information about the  probability of possible distributions.

**Comment by ThorvaldAagaard** [comment]
Yes, the endgame solver would receive visible cards for the player and dummy- and information about any voids, Later we could add probability for each card if known. But first we need to find the endgame solver :-)

**Comment by ThorvaldAagaard** [comment]
One idea might be to implement the anti-claim pattern - basically we give declarer worst possible hand, and if any card prevent declarer from claiming, then it is the card we want to discard

**Comment by ThorvaldAagaard** [comment]
The discard is decided by the NN because double dummy sees the same result for all of them, and the NN does not know that HT can never win. In #188. Closing.
